src/main/java/app/accommodationbookingservice/service/impl/BookingServiceImpl.java

```
@@ -0,0 +1,110 @@
+            BookingRepository repo,
+            AccommodationRepository accommodationRepo,
+            PaymentRepository paymentRepo,
+            NotificationService notificationService) {
+        this.repo = repo;
+        this.accommodationRepo = accommodationRepo;
+        this.paymentRepo = paymentRepo;
+        this.notificationService = notificationService;
+    }
+
+    @Override
+    public Booking create(Booking b) {
+        List<?> pendingPayments = paymentRepo.findPendingByUserId(b.getUserId());
+        if (!pendingPayments.isEmpty()) {
+            throw new IllegalStateException(
+                    "You have pending payments; please complete them before booking.");
+        }
+
+        List<Booking> existing = repo.findByAccommodationIdAndStatusIn(
+                b.getAccommodationId(),
+                List.of(BookingStatus.PENDING, BookingStatus.CONFIRMED)
+        );
+        boolean overlaps = existing.stream().anyMatch(ex ->
+                !(b.getCheckOutDate().isBefore(ex.getCheckInDate())
+                        || b.getCheckInDate().isAfter(ex.getCheckOutDate()))
+        );
+        if (overlaps) {
+            throw new IllegalStateException(
+                    "Accommodation " + b.getAccommodationId()
+                            + " is already booked for these dates");
+        }
+
+        Accommodation acc = accommodationRepo.findById(b.getAccommodationId()).orElseThrow();
+        if (acc.getAvailability() <= 0) {
+            throw new IllegalStateException(
+                    "No availability for accommodation " + b.getAccommodationId());
+        }
+        acc.setAvailability(acc.getAvailability() - 1);
+        accommodationRepo.save(acc);
+
+        b.setStatus(BookingStatus.PENDING);
+        Booking saved = repo.save(b);
+        notificationService.notify("New booking created: " + saved.getId());
+        return saved;
+    }
+
+    @Override
+    public Booking update(Long id, Booking b) {
```

Missing validation:

1. Should re‐run the same overlap check if dates change.
2. Should validate `checkIn < checkOut.`
3. Should throw 404 if booking not found via a custom exception with `@ResponseStatus(HttpStatus.NOT_FOUND).`